### PR TITLE
vhost-user: Add support for snapshot/restore

### PR DESCRIPTION
**Pull request by sboeuf**
Add the support for snapshot/restore to all vhost-user devices from the VMM device model. Only partial support is provided for virtio-fs since it will require some extra support from the `virtiofsd` daemon.

### CODE REVIEW

File: virtio-devices/src/vhost_user/vu_common_ctrl.rs [hunk 1]
```diff
@@ -197,7 +198,7 @@ impl VhostUserHandle {
                 .set_vring_base(
                     queue_index,
                     queue
-                        .avail_index_from_memory(mem)
+                        .used_index_from_memory(mem)
```
Comment (likebreath) [inline virtio-devices/src/vhost_user/vu_common_ctrl.rs:201]: Question: why is this change needed?

Comment (sboeuf) [reply virtio-devices/src/vhost_user/vu_common_ctrl.rs:201]: This is needed to keep the whole code in sync. In [here](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/virtio-devices/src/transport/pci_device.rs#L506-L511) we use it as well on the restore codepath.
The reason for using `used_index_from_memory()` instead of `avail_index_from_memory()` is because we want to start at the correct index based on what was done by the backend. Let's say the backend fully processed a descriptor right before the pause, it will have updated the index in the used ring. Relying on `used` is more correct than relying on `avail` since `used` is updated by the backend while `avail` is updated by the guest.

**Review by rbradford — APPROVED** [review]

**Review by likebreath — APPROVED** [review]
